File: src/algorithms/TokenBucket.js
/**
 * Token Bucket Algorithm Implementation
 *
 * Concept:
 * - Bucket has a maximum capacity of tokens
 * - Tokens are added at a constant rate (refill rate)
 * - Each request consumes tokens
 * - If not enough tokens, request is rejected
 * - Allows bursts up to bucket capacity
 *
 * Use Cases:
 * - General-purpose rate limiting
 * - Allowing burst traffic while limiting average rate
 * - API throttling with flexibility
 */

export class TokenBucket {
  constructor(capacity, refillRate) {
    if (capacity <= 0) {
      throw new Error("Capacity must be positive");
    }
    if (refillRate <= 0) {
      throw new Error("Refill rate must be positive");
    }

    this.capacity = capacity;
    this.refillRate = refillRate; // tokens per second
    this.tokens = capacity; // Start with full bucket
    this.lastRefill = Date.now();
  }

  /**
   * Try to consume tokens from the bucket
   * @param {number} tokens - Number of tokens to consume (default: 1)
   * @returns {boolean} true if request is allowed, false if rate limited
   */
  consume(tokens = 1) {
    this.refill();

    if (this.tokens >= tokens) {
      this.tokens -= tokens;
      return true;
    }

    return false;
  }

  /**
   * Refill tokens based on elapsed time
   */
  refill() {
    const now = Date.now();
    const elapsedSeconds = (now - this.lastRefill) / 1000;

    // Calculate tokens to add
    const tokensToAdd = elapsedSeconds * this.refillRate;

    // Update tokens (capped at capacity)
    this.tokens = Math.min(this.capacity, this.tokens + tokensToAdd);
    this.lastRefill = now;
  }

  /**
   * Get current number of available tokens
   * @returns {number}
   */
  getAvailableTokens() {
    this.refill();
    return Math.floor(this.tokens);
  }

  /**
   * Get time in milliseconds until bucket has at least N tokens
   * @param {number} requiredTokens
   * @returns {number}
   */
  getTimeUntilTokensAvailable(requiredTokens) {
    this.refill();

    if (this.tokens >= requiredTokens) {
      return 0;
    }

    const tokensNeeded = requiredTokens - this.tokens;
    const secondsNeeded = tokensNeeded / this.refillRate;
    return Math.ceil(secondsNeeded * 1000);
  }

  /**
   * Reset bucket to full capacity
   */
  reset() {
    this.tokens = this.capacity;
    this.lastRefill = Date.now();
  }

  /**
   * Get current state for serialization
   * @returns {{tokens: number, lastRefill: number}}
   */
  getState() {
    return {
      tokens: this.tokens,
      lastRefill: this.lastRefill,
    };
  }

  /**
   * Restore state from serialized data
   * @param {number} capacity
   * @param {number} refillRate
   * @param {{tokens: number, lastRefill: number}} state
   * @returns {TokenBucket}
   */
  static fromState(capacity, refillRate, state) {
    const bucket = new TokenBucket(capacity, refillRate);
    bucket.tokens = state.tokens;
    bucket.lastRefill = state.lastRefill;
    return bucket;
  }
}

File: src/redis/scripts/tokenBucket.lua
-- Token Bucket Algorithm (Atomic Redis Implementation)
-- KEYS[1]: Rate limit key (e.g., "ratelimit:token:user:123")
-- ARGV[1]: Capacity (maximum tokens)
-- ARGV[2]: Refill rate (tokens per second)
-- ARGV[3]: Current timestamp (milliseconds)
-- ARGV[4]: Tokens requested (default: 1)
-- ARGV[5]: TTL in seconds

local key = KEYS[1]
local capacity = tonumber(ARGV[1])
local refill_rate = tonumber(ARGV[2])
local now = tonumber(ARGV[3])
local requested = tonumber(ARGV[4]) or 1
local ttl = tonumber(ARGV[5])

-- Get current bucket state
local bucket = redis.call('HMGET', key, 'tokens', 'last_refill')
local tokens = tonumber(bucket[1])
local last_refill = tonumber(bucket[2])

-- Initialize if bucket doesn't exist
if not tokens then
  tokens = capacity
  last_refill = now
end

-- Calculate tokens to refill
local elapsed_seconds = (now - last_refill) / 1000.0
local tokens_to_add = elapsed_seconds * refill_rate

-- Refill tokens (capped at capacity)
tokens = math.min(capacity, tokens + tokens_to_add)

-- Update last refill time
last_refill = now

-- Try to consume tokens
local allowed = 0
if tokens >= requested then
  tokens = tokens - requested
  allowed = 1
end

-- Save updated state
redis.call('HMSET', key, 'tokens', tostring(tokens), 'last_refill', tostring(last_refill))
redis.call('EXPIRE', key, ttl)

-- Calculate retry after (milliseconds until requested tokens available)
local retry_after = 0
if allowed == 0 then
  local tokens_needed = requested - tokens
  retry_after = math.ceil((tokens_needed / refill_rate) * 1000)
end

-- Return: [allowed (0|1), remaining_tokens, retry_after_ms]
return {allowed, math.floor(tokens), retry_after}
